src/dataflow/analysis/flow-graph.ts

```diff
@@ -117,33 +150,90 @@ function* allInputPaths(startEdge: Edge): Iterable<BackwardsPath> {
-function evaluateCheckForPath(check: Check, path: BackwardsPath): boolean {
-  // TODO: Support boolean expression trees properly! Currently we only deal with a single string of OR'd conditions.
-  const conditions: CheckCondition[] = [];
-  switch (check.expression.type) {
-    case 'and':
-      throw new Error(`Boolean expressions with 'and' are not supported yet.`);
-    case 'or':
-      for (const child of check.expression.children) {
-        assert(child.type !== 'or' && child.type !== 'and', 'Nested boolean expressions are not supported yet.');
-        conditions.push(child as CheckCondition);
+/**
+ * Collects all the tag claims along the given path, canceling tag claims that are
+ * negated by "not" claims for the same tag downstream in the path, and ignoring
+ * "not" claims without corresponding positive claims upstream, as these dangling
+ * "not" claims are irrelevant for the given path. Note that "derives from" claims
+ * only prune paths, and are dealt with during path generation, so they are ignored.
+ */
+function computeTagClaimsInPath(path: BackwardsPath): ClaimIsTag[] {
+  const claims: ClaimIsTag[] = [];
+  // We traverse the path in the forward direction, so we can cancel correctly.
+  const edgesInPath = path.edges.slice().reverse();
+  edgesInPath.forEach(e => {
+    if (!e.claim) {
+      return;
+    }
+    if (e.claim.type === ClaimType.DerivesFrom) {
+      return;
+    } 
+    const tagClaim = e.claim as ClaimIsTag;
+    if (!tagClaim.isNot) {
+      claims.push(tagClaim);
+      return;          
+    }
+    // Our current claim is a "not" tag claim. 
+    // Ignore it if there are no preceding tag claims
+    if (claims.length === 0) {
+      return;
+    }
+    // Find and remove all existing positive claims for the tag. There may be > 1.
+    let index = 0;
+    while (index < claims.length) {
+      const priorClaim = claims[index];
+      if (priorClaim.tag == tagClaim.tag) {
+        claims.splice(index, 1);
+      } else {
+        index += 1;
       }
-      break;
-    default:
-      // Expression is just a single condition.
-      conditions.push(check.expression);
-      break;
+    }
+  });
+  return claims;
+}
+
+/**
+ * Collects and returns the names of all the handles in this path. 
+ */
+function computeHandlesInPath(path: BackwardsPath): HandleNode[] {
+  const handles: HandleNode[] = [];
+  for (const node of path.nodes) {
+    if (node instanceof HandleNode) {
+      handles.push(node);
+    }
   }
-  
-  // Check every condition against the whole path.
-  // NOTE: This is very inefficient. We check every condition against every edge in the path.
+  return handles;
+}
+
+/** 
+ * Returns true if the given check passes against one of the tag claims or one
+ * one of the handles in the path. Only one condition needs to pass.
+ */
+function evaluateCheck(conditions: CheckCondition[], 
+                       claims: ClaimIsTag[],
+                       handles: HandleNode[]): boolean {
```

Comment:
We don't seem to wrap arguments this way, as far as I've seen. Either all on one line, or each on its own line, indented 4 spaces.

Reply:
Done.